taraxa/state/dpos/precompiled/dpos_contract.go

```diff
@@ -0,0 +1,1022 @@
+
+	self.validators.Init(&self.storage, field_validators)
+	self.delegations.Init(&self.storage, field_delegations)
+	self.undelegations.Init(&self.storage, field_undelegations)
+
+	return self
+}
+
+func (self *Contract) UpdateStorage(readStorage Reader) {
+	self.delayedStorage = readStorage
+}
+
+///// DO WE NEED THIS ? ////
+func (self *Contract) UpdateConfig(cfg Config) {
+	self.cfg = cfg
+}
+
+////////////////////////////////////////////////////////////////
+
+func (self *Contract) Register(registry func(*common.Address, vm.PrecompiledContract)) {
+	defensive_copy := *contract_address
+	registry(&defensive_copy, self)
+}
+
+func (self *Contract) RequiredGas(ctx vm.CallFrame, evm *vm.EVM) uint64 {
+	// TODO: based on method being called, calculate the gas somehow. Doing it based on the length of input is totally useless
+	return uint64(len(ctx.Input)) * 20
+}
+
+func (self *Contract) lazy_init() {
+	if self.lazy_init_done {
+		return
+	}
+	self.lazy_init_done = true
+	self.storage.Get(stor_k_1(field_eligible_count), func(bytes []byte) {
+		self.eligible_count_orig = bin.DEC_b_endian_compact_64(bytes)
+	})
+	self.eligible_count = self.eligible_count_orig
+	self.storage.Get(stor_k_1(field_eligible_vote_count), func(bytes []byte) {
+		self.eligible_vote_count_orig = bin.DEC_b_endian_compact_64(bytes)
+	})
+	self.eligible_vote_count = self.eligible_vote_count_orig
+	self.amount_delegated_orig = bigutil.Big0
+	self.storage.Get(stor_k_1(field_amount_delegated), func(bytes []byte) {
+		self.amount_delegated_orig = bigutil.FromBytes(bytes)
+	})
+	self.amount_delegated = self.amount_delegated_orig
+}
+
+func (self *Contract) BeginBlockCall(rewards map[common.Address]*big.Int) {
+	for validator, reward := range rewards {
+		self.update_rewards(&validator, reward)
+	}
+}
+
+func (self *Contract) EndBlockCall() {
+	if !self.lazy_init_done {
+		return
+	}
+	// Update values
+	if self.eligible_count_orig != self.eligible_count {
+		self.storage.Put(stor_k_1(field_eligible_count), bin.ENC_b_endian_compact_64_1(self.eligible_count))
+		self.eligible_count_orig = self.eligible_count
+	}
+	if self.eligible_vote_count_orig != self.eligible_vote_count {
+		self.storage.Put(stor_k_1(field_eligible_vote_count), bin.ENC_b_endian_compact_64_1(self.eligible_vote_count))
+		self.eligible_vote_count_orig = self.eligible_vote_count
+	}
+	if self.amount_delegated_orig.Cmp(self.amount_delegated) != 0 {
+		self.storage.Put(stor_k_1(field_amount_delegated), self.amount_delegated.Bytes())
+		self.amount_delegated_orig = self.amount_delegated
+	}
+}
+
+func (self *Contract) CommitCall(readStorage Reader) {
+	defer self.storage.ClearCache()
+	// Storage Update
+	self.delayedStorage = readStorage
+}
+
+func (self *Contract) ApplyGenesis() error {
+	self.lazy_init()
+
+	for _, entry := range self.cfg.GenesisState {
+		// TODO fill this
+		var args RegisterValidatorArgs
+		self.apply_genesis_entry(&entry.Benefactor, new(big.Int).SetInt64(0), args, entry.Transfers)
```

Comment:
Do you actually need this function at all ? Why dont we call `registerValidator()` here instead ?

Reply:
yes, registervalidator use some transaction specific values. Applying genesis is not transaction (but it could be maybe)

Reply:
I think that tx specific values can be set to `var args RegisterValidatorArgs` and  tne simply call registerValidator. Maybe we can create task for that